packages/forms/src/elements/search/search.tsx

```diff
@@ -0,0 +1,148 @@
+import React, { useState, useEffect } from 'react';
+import { Field, FieldRenderProps } from 'react-final-form';
+import { Flex } from '@tpr/core';
+import { StyledInputLabel, InputElementHeading } from '../elements';
+import { FieldProps, FieldExtraProps } from '../../renderFields';
+import AccessibilityHelper from '../accessibilityHelper';
+import Autocomplete from 'accessible-autocomplete/react';
+import { filterResults, formatItemDefault } from './filterResults';
+import styles from './search.module.scss';
+
+interface SearchProps extends FieldRenderProps<string>, FieldExtraProps {
+	callback?: Function;
+	formatItem?: (item: any) => string;
+	getSelectedItem?: (item: any) => string;
+	keyValue: string;
+	notFoundMessage?: string;
+	optionsArray?: any[];
+	searchService?: (x:string) => Promise<any>;
+}
+
+type PanelVisibility = 'visible' | 'hidden' | 'complete';
+
+const Search: React.FC<SearchProps> = React.memo(
+	({
+		callback,
+		cfg,
+		formatItem = formatItemDefault,
+		getSelectedItem,
+		hint,
+		inputWidth = 10,
+		keyValue,
+		label,
+		meta,
+		name,
+		notFoundMessage = 'Your search criteria has no match',
+		optionsArray = [],
+		placeholder,
+		required = true,
+		searchService,
+		testId = 'search',
+		...rest
+	}) => {
+		const [panelVisible, setPanelVisible] = useState<PanelVisibility>('hidden');
+		const [classes, setClasses] = useState(styles.autocomplete);
+		const [optionsArrayObjects, setOptionsArrayObjects] = useState([]);
+
+		const getSelectedItemDefault = (item) => {
+			return item && item[keyValue];
+		}
+
+		const setGlobalArrays = (resultsFiltered) => {			
+			setOptionsArrayObjects(resultsFiltered);
+			if (panelVisible == 'hidden') setPanelVisible('visible');
+		}
+
+		// Function that filters the results when the options are passed directly as an array of values
+		const showResults = (query:string):any[] => {
+			const results = filterResults(query, optionsArray, keyValue);
+			setGlobalArrays(results);
+			return results;
+		};
+
+		// Function that process the results when receiving 'searchService' for making the search
+		const useSearchService = (apiResponse):any[] => {
+			setGlobalArrays(apiResponse);
+			return apiResponse;
+		}
+
+		const getResults = async (query:string, populateResults:Function) => {
```

Comment:
It's borderline, but this method isn't really presentation logic so it and it's child methods could potentially be moved into a service class to simplify and isolate the UI code (thought admittedly you'd need to pass in delegates for the state modfications).  

There's some benefit to moving the code beyond clarity since it would be easier to write targeted unit tests against the business logic (rather than the presentation logic) if that were needed in future (if a bug was found for example).